sql/snapshot.sql

```diff
@@ -0,0 +1,251 @@
+
+CREATE TABLE dag_transaction (
+	hash varchar NOT NULL,
+	block_hash varchar NOT NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL,
+	CONSTRAINT dag_transaction_pk PRIMARY KEY (hash),
+	CONSTRAINT dag_transaction_abstract_transaction_fk FOREIGN KEY (hash) REFERENCES abstract_transaction(hash),
+	CONSTRAINT dag_transaction_dag_block_fk FOREIGN KEY (block_hash) REFERENCES dag_block(hash)
+);
+
+
+-- public.fee_transaction definition
+
+-- Drop table
+
+-- DROP TABLE fee_transaction;
+
+CREATE TABLE fee_transaction (
+	hash varchar NOT NULL,
+	amount int8 NOT NULL,
+	source_addr varchar NOT NULL,
+	destination_addr varchar NOT NULL,
+	parent jsonb NOT NULL,
+	metagraph_snapshot_hash varchar NULL,
+	metagraph_id varchar NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL,
+	CONSTRAINT fee_transaction_pkey PRIMARY KEY (hash),
+	CONSTRAINT fee_transaction_metagraph_snapshot_fk FOREIGN KEY (metagraph_id,metagraph_snapshot_hash) REFERENCES metagraph_snapshot(metagraph_id,hash)
+);
+
+
+-- public.metagraph_balance_change definition
+
+-- Drop table
+
+-- DROP TABLE metagraph_balance_change;
+
+CREATE TABLE metagraph_balance_change (
+	metagraph_id varchar NOT NULL,
+	metagraph_hash varchar NOT NULL,
+	address varchar NOT NULL,
+	balance int8 NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL,
+	CONSTRAINT metagraph_balance_change_metagraph_snapshot_fk FOREIGN KEY (metagraph_id,metagraph_hash) REFERENCES metagraph_snapshot(metagraph_id,hash)
+);
+
+
+-- public.metagraph_block definition
+
+-- Drop table
+
+-- DROP TABLE metagraph_block;
+
+CREATE TABLE metagraph_block (
+	hash varchar NOT NULL,
+	metagraph_id varchar NOT NULL,
+	snapshot_hash varchar NOT NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL,
+	CONSTRAINT metagraph_block_pk PRIMARY KEY (hash),
+	CONSTRAINT metagraph_block_block_fk FOREIGN KEY (hash) REFERENCES block(hash),
+	CONSTRAINT metagraph_block_metagraph_snapshot_fk FOREIGN KEY (metagraph_id,snapshot_hash) REFERENCES metagraph_snapshot(metagraph_id,hash)
+);
+
+
+-- public.metagraph_reward_transaction definition
+
+-- Drop table
+
+-- DROP TABLE metagraph_reward_transaction;
+
+CREATE TABLE metagraph_reward_transaction (
+	destination_addr varchar NOT NULL,
+	amount int8 NOT NULL,
+	metagraph_id varchar NOT NULL,
+	metagraph_snapshot_hash varchar NOT NULL,
+	created_at timestamp DEFAULT now() NOT NULL,
+	updated_at timestamp DEFAULT now() NOT NULL,
+	CONSTRAINT metagraph_reward_transaction_metagraph_reward_transaction_fk FOREIGN KEY (metagraph_id,metagraph_snapshot_hash) REFERENCES metagraph_snapshot(metagraph_id,hash)
+);
+
+
+-- public.metagraph_snapshots_to_global_snapshot definition
+
+-- Drop table
+
+-- DROP TABLE metagraph_snapshots_to_global_snapshot;
+
+CREATE TABLE metagraph_snapshots_to_global_snapshot (
```

Comment:
I think this works - metagraph_id and metagraph_snapshot_hash have to be NOT NULL and the FK to metagraph_snapshot would need to be removed because these rows would be inserted before the snapshot records exist. This table does feel a little bit messy though if it doesn't have FK constraints. 

It might be easier to just to the `metagraph_snapshot_count` or a JSONB field with an array of metagraph snapshot records on the global snapshot. They would only be used while upload of the metagraph snapshots is pending.